File: frontend/src/api/pagination.ts
/**
 * Cursor / offset pagination helpers.
 * Backend envelope: { data: T[], pagination: { next_cursor, has_more, limit, offset?, total? } }
 *
 * Submission-service supports both cursor (next-only) and offset pagination
 * on the same endpoints. Pass ``offset`` for numbered page UIs (1 2 3 4 …),
 * ``cursor`` for infinite-scroll. ``total`` is always returned by endpoints
 * that support offset paging, so the UI can render the full page strip.
 */
import type { Paginated } from './types';

export interface ListParams {
  cursor?: string | null;
  offset?: number;
  limit?: number;
  sort?: string;
  q?: string;
}

export function buildListParams(p: ListParams): Record<string, string | number> {
  const out: Record<string, string | number> = {};
  if (p.cursor) out.cursor = p.cursor;
  if (p.offset !== undefined && p.offset > 0) out.offset = p.offset;
  if (p.limit) out.limit = p.limit;
  if (p.sort) out.sort = p.sort;
  if (p.q) out.q = p.q;
  return out;
}

/** Flatten pages from useInfiniteQuery into a single array. */
export function flattenPages<T>(pages: Paginated<T>[] | undefined): T[] {
  if (!pages) return [];
  return pages.flatMap((p) => p.data);
}

File: services/submission/src/submission_service/common/pagination.py
"""Cursor / offset pagination helpers.

Two pagination modes coexist:

- **Cursor** (legacy): opaque cursor, "next" only. Cheap, no COUNT(*).
- **Offset** (page-style): ``offset`` + ``limit`` + a `total` count. Lets the
  UI render numbered page buttons (1 2 3 4 …) like Yandex.Contest does.

Both modes share the same ``Page`` envelope; the response carries `total`
and `offset` so the client can pick whichever it cares about. Endpoints that
support offset mode populate both, endpoints that don't leave `total` as
``None``.
"""
from __future__ import annotations

import base64
import json
from typing import Any, Generic, TypeVar

from pydantic import BaseModel, Field

T = TypeVar("T")


class PageInfo(BaseModel):
    next_cursor: str | None = None
    has_more: bool = False
    limit: int = 50
    # Offset-pagination fields. Optional so existing cursor-only endpoints
    # don't have to populate them.
    offset: int = 0
    total: int | None = None


class Page(BaseModel, Generic[T]):
    data: list[T]
    pagination: PageInfo


class PageQuery(BaseModel):
    cursor: str | None = None
    offset: int = Field(default=0, ge=0)
    limit: int = Field(default=50, ge=1, le=200)
    sort: str | None = None


def encode_cursor(payload: dict[str, Any]) -> str:
    raw = json.dumps(payload, separators=(",", ":"), default=str).encode()
    return base64.urlsafe_b64encode(raw).decode().rstrip("=")


def decode_cursor(cursor: str | None) -> dict[str, Any] | None:
    if not cursor:
        return None
    try:
        padded = cursor + "=" * (-len(cursor) % 4)
        raw = base64.urlsafe_b64decode(padded.encode())
        return json.loads(raw)
    except Exception:
        return None
